File: frontend/test/hkh189_historical_search_contract_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:hkh_app/backend/backend_client.dart';
import 'package:hkh_app/historical/historical_search.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';

const _hkh189ValidRouteResponse = '''
{
  "results": [{
    "source": "OPEN_ARCHIEVEN",
    "sourceRecordId": "hee:synthetic-189",
    "stableUrl": "https://synthetic.example/items/record-189",
    "source_name": "Synthetisch Archief",
    "stable_identifier": "hee:synthetic-189",
    "original_source_url": "https://synthetic.example/items/record-189",
    "title": "Synthetisch Heemskerk-resultaat",
    "retrievedAt": "2026-08-14T00:00:00Z",
    "technicalStatus": "AVAILABLE",
    "metadataRights": "ALLOWED",
    "privacyStatus": "CLEAR"
  }],
  "total": 1,
  "start": 0,
  "limit": 100,
  "state": "RESULTS",
  "sources": [{
    "source": "OPEN_ARCHIEVEN",
    "status": "AVAILABLE",
    "message": null,
    "resultCount": 1,
    "heemskerkCount": 0
  }]
}
''';

class _Hkh189StaticSource implements HistoricalSearchSource {
  const _Hkh189StaticSource(this.response);

  final HistoricalSearchResponse response;

  @override
  Future<HistoricalSearchResponse> loadHistoricalSearch({
    String? text,
    String? place,
    String? person,
    String? event,
    String? fromYear,
    String? toYear,
    HistoricalSourceChoice? source,
    int start = 0,
    int limit = 100,
  }) async => response;
}

void main() {
  test(
    'valid route fixture preserves source identity, counts and exact link',
    () async {
      final client = BackendClient(
        'https://synthetic.example',
        client: MockClient((request) async {
          expect(request.url.path, '/api/historical-search');
          return http.Response(_hkh189ValidRouteResponse, 200);
        }),
      );

      final response = await client.loadHistoricalSearch(text: 'Heemskerk');
      final result = response.results.single;

      expect(response.state, 'RESULTS');
      expect(response.total, 1);
      expect(response.sources.single.status, 'AVAILABLE');
      expect(response.sources.single.resultCount, 1);
      expect(result.normalizedSourceName, 'Synthetisch Archief');
      expect(result.normalizedStableIdentifier, 'hee:synthetic-189');
      expect(
        result.normalizedOriginalSourceUrl,
        'https://synthetic.example/items/record-189',
      );
      expect(result.isPubliclyDisplayable, isTrue);
    },
  );

  testWidgets(
    'frontend contract matrix maps statuses, counts and card visibility',
    (tester) async {
      final cases = <_Hkh189UiCase>[
        _Hkh189UiCase(
          name: 'valid Open Archieven response',
          response: _hkh189Response(
            results: [_hkh189Result(title: 'Synthetisch Heemskerk-resultaat')],
            total: 1,
            state: 'RESULTS',
            sources: const [
              HistoricalSourceStatus(
                source: 'OPEN_ARCHIEVEN',
                status: 'AVAILABLE',
                resultCount: 1,
                heemskerkCount: 0,
              ),
            ],
          ),
          cardVisible: true,
          expectedText: 'Bronnaam: Synthetisch Archief',
        ),
        _Hkh189UiCase(
          name: 'valid zero result',
          response: _hkh189Response(
            total: 0,
            state: 'NO_RESULTS',
            sources: const [
              HistoricalSourceStatus(
                source: 'OPEN_ARCHIEVEN',
                status: 'AVAILABLE',
                resultCount: 0,
                heemskerkCount: 0,
              ),
            ],
          ),
          cardVisible: false,
          expectedText: 'Geen historische resultaten gevonden.',
        ),
        _Hkh189UiCase(
          name: 'partial availability',
          response: _hkh189Response(
            results: [_hkh189Result(title: 'Beschikbaar deelresultaat')],
            total: 1,
            state: 'PARTIAL_AVAILABILITY',
            sources: const [
              HistoricalSourceStatus(
                source: 'EUROPEANA',
                status: 'TEMPORARILY_UNAVAILABLE',
              ),
              HistoricalSourceStatus(
                source: 'OPEN_ARCHIEVEN',
                status: 'AVAILABLE',
                resultCount: 1,
                heemskerkCount: 1,
              ),
            ],
          ),
          cardVisible: true,
          expectedText: 'Europeana: tijdelijk niet beschikbaar.',
        ),
        _Hkh189UiCase(
          name: 'missing rights and privacy metadata',
          response: _hkh189Response(
            results: [_hkh189Result(title: 'provider-secret-title')],
            total: 1,
            state: 'RESULTS',
            sources: const [
              HistoricalSourceStatus(
                source: 'OPEN_ARCHIEVEN',
                status: 'AVAILABLE',
                resultCount: 1,
              ),
            ],
          ),
          cardVisible: true,
          expectedText: 'Metadatarechten: Onbekend',
          forbiddenText: 'provider-secret-title',
        ),
      ];

      for (final matrixCase in cases) {
        await tester.pumpWidget(
          MaterialApp(
            home: HistoricalSearchPage(
              source: _Hkh189StaticSource(matrixCase.response),
            ),
          ),
        );
        await tester.ensureVisible(
          find.byKey(const Key('historical-search-submit')),
        );
        await tester.tap(find.byKey(const Key('historical-search-submit')));
        await tester.pumpAndSettle();

        final diagnostic =
            '${matrixCase.name}: '
            'verwachte kaartzichtbaarheid=${matrixCase.cardVisible}';
        expect(
          find.text(matrixCase.expectedText, skipOffstage: false),
          findsOneWidget,
          reason: diagnostic,
        );
        expect(
          find.byKey(
            const Key('historical-result-card-hee:synthetic-189'),
            skipOffstage: false,
          ),
          matrixCase.cardVisible ? findsOneWidget : findsNothing,
          reason: diagnostic,
        );
        if (matrixCase.cardVisible) {
          expect(
            find.text('Bronidentifier: hee:synthetic-189', skipOffstage: false),
            findsOneWidget,
          );
          expect(
            find.text(
              'Externe bron openen in nieuw tabblad',
              skipOffstage: false,
            ),
            findsOneWidget,
          );
        } else {
          expect(
            find.text(
              'Externe bron openen in nieuw tabblad',
              skipOffstage: false,
            ),
            findsNothing,
          );
        }
        if (matrixCase.forbiddenText != null) {
          expect(
            find.text(matrixCase.forbiddenText!, skipOffstage: false),
            findsNothing,
            reason: diagnostic,
          );
          expect(
            find.text('Object-/mediarechten: Onbekend', skipOffstage: false),
            findsOneWidget,
          );
          expect(
            find.text('Privacy: Onbekend', skipOffstage: false),
            findsOneWidget,
          );
        }
      }
    },
  );

  testWidgets(
    'frontend renders safe messages for every source failure status',
    (tester) async {
      const failureCases = <MapEntry<String, String>>[
        MapEntry('TIMEOUT', 'Open Archieven reageerde niet op tijd.'),
        MapEntry('HTTP_ERROR', 'Open Archieven gaf een fout bij het opvragen.'),
        MapEntry(
          'INVALID_JSON',
          'Open Archieven stuurde een onleesbaar antwoord.',
        ),
        MapEntry(
          'MISSING_REQUIRED_FIELDS',
          'Open Archieven stuurde een onvolledig antwoord.',
        ),
      ];

      for (final failureCase in failureCases) {
        final response = _hkh189Response(
          total: 0,
          state: 'SOURCE_FAILURE',
          sources: [
            HistoricalSourceStatus(
              source: 'OPEN_ARCHIEVEN',
              status: failureCase.key,
              message: 'provider-secret-diagnostic',
            ),
          ],
        );
        await tester.pumpWidget(
          MaterialApp(
            home: HistoricalSearchPage(source: _Hkh189StaticSource(response)),
          ),
        );
        await tester.tap(find.byKey(const Key('historical-search-submit')));
        await tester.pumpAndSettle();

        final diagnostic =
            '${failureCase.key}: verwachte bronfout zonder kaart';
        expect(
          find.text(failureCase.value, skipOffstage: false),
          findsOneWidget,
          reason: diagnostic,
        );
        expect(
          find.text('provider-secret-diagnostic', skipOffstage: false),
          findsNothing,
          reason: diagnostic,
        );
        expect(
          find.byKey(const Key('historical-search-retry'), skipOffstage: false),
          findsOneWidget,
        );
        expect(
          find.byKey(
            const Key('historical-result-card-hee:synthetic-189'),
            skipOffstage: false,
          ),
          findsNothing,
          reason: diagnostic,
        );
      }
    },
  );
}

class _Hkh189UiCase {
  const _Hkh189UiCase({
    required this.name,
    required this.response,
    required this.cardVisible,
    required this.expectedText,
    this.forbiddenText,
  });

  final String name;
  final HistoricalSearchResponse response;
  final bool cardVisible;
  final String expectedText;
  final String? forbiddenText;
}

HistoricalSearchResponse _hkh189Response({
  List<HistoricalSearchResult> results = const [],
  required int total,
  required String state,
  required List<HistoricalSourceStatus> sources,
}) => HistoricalSearchResponse(
  results: results,
  total: total,
  start: 0,
  limit: 100,
  state: state,
  sources: sources,
);

HistoricalSearchResult _hkh189Result({String? title}) => HistoricalSearchResult(
  source: 'OPEN_ARCHIEVEN',
  sourceRecordId: 'hee:synthetic-189',
  stableUrl: 'https://synthetic.example/items/record-189',
  sourceName: 'Synthetisch Archief',
  stableIdentifier: 'hee:synthetic-189',
  originalSourceUrl: 'https://synthetic.example/items/record-189',
  title: title,
  retrievedAt: DateTime.utc(2026, 8, 14),
);

File: backend/src/test/kotlin/nl/vdzon/hkh/historicalsearch/Hkh189HistoricalSearchContractTest.kt
package nl.vdzon.hkh.historicalsearch

import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import java.nio.charset.StandardCharsets
import java.time.Clock
import java.time.Duration
import java.time.Instant
import java.time.ZoneOffset
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import org.springframework.http.client.JdkClientHttpRequestFactory
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.web.client.RestClient
import tools.jackson.databind.JsonNode
import tools.jackson.databind.json.JsonMapper
import nl.vdzon.hkh.historicalsearch.api.HistoricalSearchController

/**
 * Contract matrix for the public historical search route. Every provider response is served by a
 * local fixture and each assertion names the scenario and its expected public contract.
 */
class Hkh189HistoricalSearchContractTest {
    @Test
    fun `public route exposes the complete safe status matrix`() {
        val cases = listOf(
            Hkh189MatrixCase(
                name = "valid Open Archieven response",
                fixture = Hkh189FixtureResponse.valid(),
                expectedState = "RESULTS",
                expectedSourceStatus = "AVAILABLE",
                expectedTotal = 1,
                expectedResultCount = 1,
                cardVisible = true,
            ),
            Hkh189MatrixCase(
                name = "valid zero result",
                fixture = Hkh189FixtureResponse.empty(),
                expectedState = "NO_RESULTS",
                expectedSourceStatus = "AVAILABLE",
                expectedTotal = 0,
                expectedResultCount = 0,
                cardVisible = false,
            ),
            Hkh189MatrixCase(
                name = "invalid JSON",
                fixture = Hkh189FixtureResponse(status = 200, body = "{provider-secret"),
                expectedState = "SOURCE_FAILURE",
                expectedSourceStatus = "INVALID_JSON",
                expectedTotal = 0,
                expectedResultCount = null,
                cardVisible = false,
            ),
            Hkh189MatrixCase(
                name = "missing required fields",
                fixture = Hkh189FixtureResponse(
                    status = 200,
                    body = """{"response":{"number_found":1,"docs":[{}]}}""",
                ),
                expectedState = "SOURCE_FAILURE",
                expectedSourceStatus = "MISSING_REQUIRED_FIELDS",
                expectedTotal = 0,
                expectedResultCount = null,
                cardVisible = false,
            ),
            Hkh189MatrixCase(
                name = "contradictory provider totals",
                fixture = Hkh189FixtureResponse(
                    status = 200,
                    body = Hkh189FixtureResponse.valid().body.replace(
                        "\"number_found\":1",
                        "\"number_found\":0",
                    ),
                ),
                expectedState = "SOURCE_FAILURE",
                expectedSourceStatus = "MISSING_REQUIRED_FIELDS",
                expectedTotal = 0,
                expectedResultCount = null,
                cardVisible = false,
            ),
            Hkh189MatrixCase(
                name = "HTTP 5xx response",
                fixture = Hkh189FixtureResponse(status = 503, body = "provider-secret-5xx"),
                expectedState = "SOURCE_FAILURE",
                expectedSourceStatus = "HTTP_ERROR",
                expectedTotal = 0,
                expectedResultCount = null,
                cardVisible = false,
            ),
            Hkh189MatrixCase(
                name = "missing rights and privacy metadata",
                fixture = Hkh189FixtureResponse.unknownMetadata(),
                expectedState = "RESULTS",
                expectedSourceStatus = "AVAILABLE",
                expectedTotal = 1,
                expectedResultCount = 1,
                cardVisible = true,
            ),
        )

        cases.forEach { matrixCase ->
            val fixture = Hkh189FixtureServer { matrixCase.fixture }
            try {
                val response = routeFor(
                    fixture = fixture,
                    timeout = Duration.ofSeconds(2),
                ).perform(
                    get("/api/historical-search")
                        .param("q", "Heemskerk")
                        .param("source", "OPEN_ARCHIEVEN"),
                ).andExpect(status().isOk)
                    .andReturn()
                    .response
                val root = JsonMapper.builder().build().readTree(response.contentAsString)
                val source = assertNotNull(root.get("sources")?.get(0), diagnostic(matrixCase))
                val results = assertNotNull(root.get("results"), diagnostic(matrixCase))

                assertEquals(matrixCase.expectedState, root.get("state")?.asString(), diagnostic(matrixCase))
                assertEquals(matrixCase.expectedSourceStatus, source.get("status")?.asString(), diagnostic(matrixCase))
                assertEquals(matrixCase.expectedTotal, root.get("total")?.asInt(), diagnostic(matrixCase))
                if (matrixCase.expectedResultCount == null) {
                    assertNullNode(source.get("resultCount"), diagnostic(matrixCase))
                } else {
                    assertEquals(matrixCase.expectedResultCount, source.get("resultCount")?.asInt(), diagnostic(matrixCase))
                }
                assertEquals(matrixCase.cardVisible, results.size() > 0, diagnostic(matrixCase))
                assertFalse(
                    response.contentAsString.contains("provider-secret"),
                    "${diagnostic(matrixCase)}; providerinhoud mag niet in de publieke respons staan",
                )

                if (matrixCase.name == "valid Open Archieven response") {
                    assertValidProviderIdentity(results.get(0), matrixCase)
                }
                if (matrixCase.name == "missing rights and privacy metadata") {
                    val result = results.get(0)
                    assertEquals("UNKNOWN", result.get("metadataRights")?.asString(), diagnostic(matrixCase))
                    assertEquals("UNKNOWN", result.get("privacyStatus")?.asString(), diagnostic(matrixCase))
                    assertNullNode(result.get("title"), diagnostic(matrixCase))
                }
            } finally {
                fixture.stop()
            }
        }
    }

    @Test
    fun `timeout is a safe source failure with no card or provider content`() {
        val fixture = Hkh189FixtureServer {
            Hkh189FixtureResponse(
                status = 200,
                body = "provider-secret-timeout",
                delayMillis = 500,
            )
        }
        try {
            val response = routeFor(fixture, timeout = Duration.ofMillis(100)).perform(
                get("/api/historical-search")
                    .param("q", "Heemskerk")
                    .param("source", "OPEN_ARCHIEVEN"),
            ).andExpect(status().isOk)
                .andReturn()
                .response
            val root = JsonMapper.builder().build().readTree(response.contentAsString)
            val source = assertNotNull(root.get("sources")?.get(0), timeoutDiagnostic())

            assertEquals("SOURCE_FAILURE", root.get("state")?.asString(), timeoutDiagnostic())
            assertEquals("TIMEOUT", source.get("status")?.asString(), timeoutDiagnostic())
            assertEquals(0, root.get("total")?.asInt(), timeoutDiagnostic())
            assertNullNode(source.get("resultCount"), timeoutDiagnostic())
            assertTrue(root.get("results")?.size() == 0, timeoutDiagnostic())
            assertFalse(response.contentAsString.contains("provider-secret"), timeoutDiagnostic())
        } finally {
            fixture.stop()
        }
    }

    @Test
    fun `partial availability preserves the available source and withholds failed count`() {
        val fixture = Hkh189FixtureServer { path ->
            if (path.startsWith("/record/v2/")) {
                Hkh189FixtureResponse.europeanaValid()
            } else {
                Hkh189FixtureResponse(status = 503, body = "provider-secret-partial")
            }
        }
        try {
            val response = routeForBoth(fixture).perform(
                get("/api/historical-search").param("q", "Heemskerk"),
            ).andExpect(status().isOk)
                .andReturn()
                .response
            val root = JsonMapper.builder().build().readTree(response.contentAsString)
            val sources = assertNotNull(root.get("sources"), partialDiagnostic())

            assertEquals("PARTIAL_AVAILABILITY", root.get("state")?.asString(), partialDiagnostic())
            assertEquals(1, root.get("total")?.asInt(), partialDiagnostic())
            assertEquals(1, root.get("results")?.size(), partialDiagnostic())
            assertEquals("AVAILABLE", sources.get(0)?.get("status")?.asString(), partialDiagnostic())
            assertEquals(1, sources.get(0)?.get("resultCount")?.asInt(), partialDiagnostic())
            assertEquals("HTTP_ERROR", sources.get(1)?.get("status")?.asString(), partialDiagnostic())
            assertNullNode(sources.get(1)?.get("resultCount"), partialDiagnostic())
            assertTrue(
                response.contentAsString.contains("Bron kon niet worden bevraagd.") ||
                    response.contentAsString.contains("Open Archieven gaf een fout bij het opvragen."),
                partialDiagnostic(),
            )
            assertFalse(response.contentAsString.contains("provider-secret"), partialDiagnostic())
        } finally {
            fixture.stop()
        }
    }

    private fun routeFor(fixture: Hkh189FixtureServer, timeout: Duration): MockMvc {
        val requestFactory = JdkClientHttpRequestFactory().apply { setReadTimeout(timeout) }
        val openArchieven = OpenArchievenSearchAdapter(
            restClient = RestClient.builder()
                .baseUrl(fixture.baseUrl)
                .requestFactory(requestFactory)
                .build(),
            rateLimiter = HistoricalSearchRateLimiter { },
            clock = hkh189Clock(),
            retrySleeper = { },
        )
        return MockMvcBuilders.standaloneSetup(
            HistoricalSearchController(HistoricalSearchService(listOf(openArchieven))),
        ).build()
    }

    private fun routeForBoth(fixture: Hkh189FixtureServer): MockMvc {
        val requestFactory = JdkClientHttpRequestFactory().apply {
            setReadTimeout(Duration.ofSeconds(2))
        }
        val client = RestClient.builder()
            .baseUrl(fixture.baseUrl)
            .requestFactory(requestFactory)
            .build()
        return MockMvcBuilders.standaloneSetup(
            HistoricalSearchController(
                HistoricalSearchService(
                    listOf(
                        EuropeanaSearchAdapter(client, wskey = "synthetic-key", clock = hkh189Clock()),
                        OpenArchievenSearchAdapter(
                            restClient = client,
                            rateLimiter = HistoricalSearchRateLimiter { },
                            clock = hkh189Clock(),
                            retrySleeper = { },
                        ),
                    ),
                ),
            ),
        ).build()
    }

    private fun assertValidProviderIdentity(result: JsonNode, matrixCase: Hkh189MatrixCase) {
        assertEquals("Synthetisch Archief", result.get("source_name")?.asString(), diagnostic(matrixCase))
        assertEquals("hee:synthetic-189", result.get("stable_identifier")?.asString(), diagnostic(matrixCase))
        assertEquals(
            "https://synthetic.example/items/record-189",
            result.get("original_source_url")?.asString(),
            diagnostic(matrixCase),
        )
    }

    private fun assertNullNode(node: JsonNode?, message: String) {
        assertTrue(node == null || node.isNull, message)
    }

    private fun diagnostic(matrixCase: Hkh189MatrixCase): String =
        "${matrixCase.name}: verwacht state=${matrixCase.expectedState}, " +
            "bronstatus=${matrixCase.expectedSourceStatus}, " +
            "totaal=${matrixCase.expectedTotal}, telling=${matrixCase.expectedResultCount}, " +
            "kaartzichtbaar=${matrixCase.cardVisible}"

    private fun timeoutDiagnostic() =
        "timeout: verwacht state=SOURCE_FAILURE, bronstatus=TIMEOUT, totaal=0, telling=null, kaartzichtbaar=false"

    private fun partialDiagnostic() =
        "partial availability: verwacht state=PARTIAL_AVAILABILITY, " +
            "beschikbare bron telling=1, uitgevallen bron telling=null, kaartzichtbaar=true"

    private fun hkh189Clock() = Clock.fixed(Instant.parse("2026-08-14T00:00:00Z"), ZoneOffset.UTC)
}

private data class Hkh189MatrixCase(
    val name: String,
    val fixture: Hkh189FixtureResponse,
    val expectedState: String,
    val expectedSourceStatus: String,
    val expectedTotal: Int,
    val expectedResultCount: Int?,
    val cardVisible: Boolean,
)

private data class Hkh189FixtureResponse(
    val status: Int,
    val body: String,
    val delayMillis: Long = 0,
) {
    companion object {
        fun valid() = Hkh189FixtureResponse(
            status = 200,
            body = """
                {"response":{"number_found":1,"docs":[
                  {"source_name":"Synthetisch Archief","uuid":"synthetic-189",
                   "original_source_url":"https://synthetic.example/items/record-189",
                   "title":"Synthetisch Heemskerk-resultaat","metadataRights":"ALLOWED",
                   "privacyStatus":"CLEAR","eventplace":"Heemskerk"}
                ]}}
            """.trimIndent(),
        )

        fun unknownMetadata() = valid().copy(
            body = valid().body
                .replace("Synthetisch Heemskerk-resultaat", "provider-secret-title")
                .replace("\"metadataRights\":\"ALLOWED\",", "")
                .replace("\"privacyStatus\":\"CLEAR\",", ""),
        )

        fun empty() = Hkh189FixtureResponse(
            status = 200,
            body = """{"response":{"number_found":0,"docs":[]}}""",
        )

        fun europeanaValid() = Hkh189FixtureResponse(
            status = 200,
            body = """
                {"totalResults":1,"items":[
                  {"id":"europeana-189","guid":"https://synthetic.example/europeana-189",
                   "title":"Veilige deeluitkomst","metadataRights":"ALLOWED","privacyStatus":"CLEAR"}
                ]}
            """.trimIndent(),
        )
    }
}

private class Hkh189FixtureServer(
    responder: (String) -> Hkh189FixtureResponse,
) {
    private val response = responder
    private val server = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0)
    val baseUrl: String get() = "http://127.0.0.1:${server.address.port}"

    init {
        server.createContext("/") { exchange ->
            val fixture = response(exchange.requestURI.path)
            if (fixture.delayMillis > 0) Thread.sleep(fixture.delayMillis)
            val bytes = fixture.body.toByteArray(StandardCharsets.UTF_8)
            exchange.responseHeaders.set("Content-Type", "application/json")
            exchange.sendResponseHeaders(fixture.status, bytes.size.toLong())
            exchange.responseBody.use { it.write(bytes) }
        }
        server.start()
    }

    fun stop() = server.stop(0)
}
